src/useBattery.ts

```diff
@@ -1,56 +1,106 @@
-import { useEffect, useState } from 'react';
+import * as React from 'react';
 import { off, on } from './util';
 
-export interface BatterySensorState {
+enum BatteryManagerEvents {
+  levelChange = 'levelchange',
+  dischargingTimeChange = 'dischargingtimechange',
+  chargingTimeChange = 'chargingtimechange',
+  chargingChange = 'chargingchange',
+}
+
+export interface BatteryState {
   charging: boolean;
-  level: number;
   chargingTime: number;
   dischargingTime: number;
+  level: number;
 }
 
-const useBattery = () => {
-  const [state, setState] = useState({});
-  let mounted = true;
-  let battery: any = null;
-
-  const onChange = () => {
-    const { charging, level, chargingTime, dischargingTime } = battery;
-    setState({
-      charging,
-      level,
-      chargingTime,
-      dischargingTime,
-    });
-  };
-
-  const onBattery = () => {
-    onChange();
-    on(battery, 'chargingchange', onChange);
-    on(battery, 'levelchange', onChange);
-    on(battery, 'chargingtimechange', onChange);
-    on(battery, 'dischargingtimechange', onChange);
-  };
-
-  useEffect(() => {
-    (navigator as any).getBattery().then((bat: any) => {
-      if (mounted) {
-        battery = bat;
-        onBattery();
-      }
-    });
-
-    return () => {
-      mounted = false;
-      if (battery) {
-        off(battery, 'chargingchange', onChange);
-        off(battery, 'levelchange', onChange);
-        off(battery, 'chargingtimechange', onChange);
-        off(battery, 'dischargingtimechange', onChange);
-      }
-    };
-  }, []);
+interface BatteryManager extends Readonly<BatteryState>, EventTarget {
+  onchargingchange: () => void;
+  onchargingtimechange: () => void;
+  ondischargingtimechange: () => void;
+  onlevelchange: () => void;
+}
 
-  return state;
+interface NavigatorWithPossibleBattery extends Navigator {
+  getBattery?: () => Promise<BatteryManager>;
+}
+
+const nav: NavigatorWithPossibleBattery = navigator;
+
+type UseBatteryState = BatteryState & {
+  isSupported: boolean;
 };
 
-export default useBattery;
+export default function useBattery() {
+  const [state, setState] = React.useState<UseBatteryState>({
+    isSupported: nav && typeof nav.getBattery !== 'undefined',
+    level: 1,
+    charging: true,
+    dischargingTime: Infinity,
+    chargingTime: 0,
+  });
+  const battery = React.useRef<BatteryManager>();
+  let isMounted = true;
+
+  if (state.isSupported) {
```

Comment:
Instead of having this `if` statement, can we instead write this hook assuming that it is client side and Battery API is supported, and then at module scope make a check.

```
const useBattery = isClient && isSupported ? useBatteryClient : useBatteryMock;
export default useBattery;
```

Reply:
Well, this one is really good idea =)